### PR TITLE
Askaimodel

### PR DESCRIPTION
**Pull request by jpdevhub**
Introduces a zero-key "Ask AI" banner functionality that appears after a user reviews an answer. It allows users to automatically compose a structured prompt containing the question context (with parsed sub-options) and deep-link instantly into external models like ChatGPT, Claude, or Grok.

Key technical implementations:

Added a Settings selector allowing users to configure their preferred deep-link destination.
Implemented full ClipboardItem fetching mechanics to transparently copy embedded question diagrams to the clipboard so users can seamlessly hit Cmd+V in the external AI composer.
Built graceful fallbacks for URL-length thresholds, multiple-image detections, and unsupported clipboard APIs.
Included full Vitest coverage for all edge-cases via aiPromptUtils.test.ts.

### CODE REVIEW

File: src/components/QuestionCard/ActionButtons.tsx [hunk 1]
```diff
@@ -10,7 +10,9 @@ type ActionButtonsProps = {
     handleShowAnswer?: (() => void) | undefined;
     handleSubmit?: (() => void) | undefined;
     handleExplainationClick: () => void;
-    isCompatible: boolean | undefined;
+    handleGenerateAIAnswer?: () => void;
+    isGeneratingAI?: boolean;
+    hasAIAnswer?: boolean;
```
Comment (Razen04) [inline src/components/QuestionCard/ActionButtons.tsx]: Remove these as I think these belong to the different PR.

File: src/components/QuestionCard/ActionButtons.tsx [hunk 2]
```diff
@@ -22,7 +24,9 @@ const ActionButtons = ({
     handleShowAnswer,
     handleSubmit,
     handleExplainationClick,
-    isCompatible,
+    handleGenerateAIAnswer,
+    isGeneratingAI,
+    hasAIAnswer,
```
Comment (Razen04) [inline src/components/QuestionCard/ActionButtons.tsx]: Same goes for these.

File: src/components/QuestionCard/ActionButtons.tsx [hunk 3]
```diff
@@ -80,9 +82,11 @@ const ActionButtons = ({
                     >
                         <ChatCircle className="inline text-lg" />
                         <span className="hidden md:inline ml-2">
-                            Show Explanation <span className="font-mono">[/]</span>
+                            Explanation <span className="font-mono">[/]</span>
```
Comment (Razen04) [inline src/components/QuestionCard/ActionButtons.tsx]: Why changed it? Any particular reason?

File: src/components/QuestionCard/AskAIBanner.tsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
... 0 0-6.51-2.9A6.065 6.065 0 0 0 4.981 4.18a5.985 5.985 0 0 0-3.998 2.9 6.046 6.046 0 0 0 .743 7.097 5.98 5.98 0 0 0 .51 4.911 6.051 6.051 0 0 0 6.515 2.9A5.985 5.985 0 0 0 13.26 24a6.056 6.056 0 0 0 5.772-4.206 5.99 5.99 0 0 0 3.997-2.9 6.056 6.056 0 0 0-.747-7.073zM13.26 22.43a4.476 4.476 0 0 1-2.876-1.04l.141-.081 4.779-2.758a.795.795 0 0 0 .392-.681v-6.737l2.02 1.168a.071.071 0 0 1 .038.052v5.583a4.504 4.504 0 0 1-4.494 4.494zM3.6 18.304a4.47 4.47 0 0 1-.535-3.014l.142.085 4.783 2.759a.771.771 0 0 0 .78 0l5.843-3.369v2.332a.08.08 0 0 1-.033.062L9.74 19.95a4.5 4.5 0 0 1-6.14-1.646zM2.34 7.896a4.485 4.485 0 0 1 2.366-1.973V11.6a.766.766 0 0 0 .388.676l5.815 3.355-2.02 1.168a.076.076 0 0 1-.071 0l-4.83-2.786A4.504 4.504 0 0 1 2.34 7.872zm16.597 3.855l-5.833-3.387 2.02-1.168a.076.076 0 0 1 .071 0l4.83 2.791a4.494 4.494 0 0 1-.676 8.105v-5.678a.79.79 0 0 0-.412-.663zm2.01-3.023l-.141-.085-4.774-2.782a.776.776 0 0 0-.785 0L9.409 9.23V6.897a.066.066 0 0 1 .033-.062l4.83-2.787a4.5 4.5 0 0 1 6.68 4.66zm-12.64 4.135l-2.02-1.164a.08.08 0 0 1-.038-.057V6.075a4.5 4.5 0 0 1 7.375-3.453l-.142.08L8.704 5.46a.795.795 0 0 0-.393.681zm1.097-2.365l2.602-1.5 2.607 1.5v2.999l-2.597 1.5-2.607-1.5z" />
```
Comment (Razen04) [inline src/components/QuestionCard/AskAIBanner.tsx]: Put these svgs inside public folder under ai_providers, i.e. `public/ai_providers/chatgpt.svg`

This way the code organisation is better.

Comment (Razen04) [reply src/components/QuestionCard/AskAIBanner.tsx]: Same for other providers too.

File: src/components/QuestionCard/AskAIBanner.tsx [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            </svg>
+        ),
+    },
+};
+
+// ---------------------------------------------------------------------------
+// Component
+// ---------------------------------------------------------------------------
+const AskAIBanner: React.FC<AskAIBannerProps> = ({ provider, onClick }) => {
+    // Fallback to chatgpt if the provider is no longer supported (e.g. users who had gemini saved)
```
Comment (Razen04) [inline src/components/QuestionCard/AskAIBanner.tsx:50]: Why no added Gemini, any particular reason? I think students have Gemini Pro so wouldn't that would be better to have too?

File: src/components/QuestionCard/QuestionCard.tsx [hunk 6]
```diff
@@ -224,8 +210,11 @@ const QuestionCard = ({
                         />
                     )}
 
-                    {/* Question Explanation */}
-                    {showAnswer && <QuestionExplanation question={question} />}
+                    {/* Question Explanation (AI cached answer) */}
+                    {showAnswer && <QuestionExplanation question={question} customAiAnswer={question.answer_text} />}
```
Comment (Razen04) [inline src/components/QuestionCard/QuestionCard.tsx]: I think this is from other PR too.

File: src/components/QuestionCard/QuestionExplanation.tsx
Comment (Razen04) [inline src/components/QuestionCard/QuestionExplanation.tsx:1]: Is this from this PR? Cause I think customAiAnswer isn't being added in this PR.

File: src/context/AppProvider.tsx [hunk 7]
```diff
@@ -34,6 +35,14 @@ const AppProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
...
     };
 
+    // A generic function to set any setting to a specific value (for non-boolean settings).
+    const handleSettingChange = <K extends keyof Settings>(key: K, value: Settings[K]) => {
+        setSettings((prev) => ({
+            ...prev,
+            [key]: value,
+        }));
+    };
+
```
Comment (Razen04) [inline src/context/AppProvider.tsx]: There is a function already called `handleSettingsToggle` which does the same work with different input, I think you should modify that instead of creating new function cause that will create ambiguity issues like which does what. 

It will be easy just check if the value is provided then it is for value change otherwise it is only a boolean toggle.

File: src/context/AppSettingContext.ts [hunk 8]
```diff
@@ -5,6 +5,7 @@ type AppSettingContextType =
     | {
           settings: Settings;
           handleSettingToggle: (key: keyof Settings) => void;
+          handleSettingChange: <K extends keyof Settings>(key: K, value: Settings[K]) => void;
```
Comment (Razen04) [inline src/context/AppSettingContext.ts]: Once you do that you won't need this.

File: src/utils/aiPromptUtils.ts [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+/**
+ * Resolve the correct answer to a human-readable string.
+ * Handles MCQ (index array), MSQ (index array), and Numerical.
+ */
+function resolveCorrectAnswer(question: Question): string {
+    const labels = ['A', 'B', 'C', 'D', 'E', 'F'];
+
+    if (question.question_type === 'Numerical Answer') {
```
Comment (Razen04) [inline src/utils/aiPromptUtils.ts]: The type is `numerical` | `multiple-choice` | `multiple-select`

File: src/utils/aiPromptUtils.ts [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+        if (ca.type === 'exact') return String(ca.value);
+        if (ca.type === 'multiple') return ca.values.join(', ');
+        if (ca.type === 'range') return `between ${ca.min} and ${ca.max}`;
+        if (ca.type === 'tolerance') return `${ca.value} ± ${ca.tolerance}`;
+        return 'See solution';
+    }
+
+    // MCQ or MSQ — correct_answer is an array of option indices
+    const q = question as MCQQuestion | MSQQuestion;
+    if (!q.options || !Array.isArray(q.correct_answer)) return 'See solution';
```
Comment (Razen04) [inline src/utils/aiPromptUtils.ts]: Let us switch too question_type now everywhere instead of this cause I am switching to it.

File: src/utils/aiPromptUtils.ts [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+
+/**
+ * Builds a rich, structured prompt optimised for GATE exam explanations.
+ * Works with any AI provider (ChatGPT, Gemini, Claude, Grok).
+ *
+ * @param question    The GATE question object.
+ * @param imageCount  The number of diagrams in the question. Used to format placeholders.
+ */
+export function buildGateAIPrompt(question: Question, imageCount = 0): string {
+    const isMCQ = question.question_type !== 'Numerical Answer';
```
Comment (Razen04) [inline src/utils/aiPromptUtils.ts]: Apply same here. just check for `multiple-choice` in question_type

**Review by Razen04 — CHANGES_REQUESTED** [review]
The image is not getting copied in the clipboard, also we should tell the user beforehand that they will have to paste the image manually before redirecting to the any AI. So will have to look into this.

File: src/components/QuestionCard/AskAIBanner.tsx (hunk 5 above)
Comment (jpdevhub) [reply src/components/QuestionCard/AskAIBanner.tsx:50]: gemini dont support the copied text from another website.

**Comment by jpdevhub** [comment]
fixes all the different pr code changes that you have asked, please look into it @Razen04

File: src/__tests__/hooks/test-engine/useTestAnswer.test.ts
Comment (Razen04) [inline src/__tests__/hooks/test-engine/useTestAnswer.test.ts:1]: Why changed these any particular reason? I don't think these test files are required and code change in this PR.

File: supabase/functions/deno.json
Comment (Razen04) [inline supabase/functions/deno.json:1]: All the files inside the supabase folder belongs to different PR, I believe, please check once.

**Comment by Razen04** [comment]
`package.json`, `tsconfig`, `vitest` and all those are from different PR's. This PR dosen't need those.

**Comment by Razen04** [comment]
Just remove those and we are good to go. 


Also was thinking about the doubt feature we discussed and also custom prompts, basically letting users change the prompt which we provide by default. 


If you can, then do it in this PR otherwise I will look into it later in the next version.

File: src/__tests__/hooks/test-engine/useTestAnswer.test.ts
Comment (jpdevhub) [reply src/__tests__/hooks/test-engine/useTestAnswer.test.ts:1]: i have just put all the test files at  one folder.

**Comment by Razen04** [comment]
I am closing the PR, I have created a follow-up PR: #65 , I will be fixing some things and build on top of it. Thanks for the contribution.
